docs/xml.md

```diff
@@ -1,6 +1,659 @@
 # Roblox XML Model Format, Version 4
 This is unofficial documentation for Roblox's XML model format. The XML model format is used for places (`.rbxlx` files), models (`.rbxmx` files), Roblox Studio settings, and many objects uploaded to Roblox's asset storage.
 
-The XML model format has generally been replaced by the newer, more efficient [binary model format](/binary). Some use cases for the XML format still exist, owing to its human readability.
+The XML model format has generally been replaced by the newer, more efficient [binary model format](binary.md). Some use cases for the XML format still exist, owing to its human readability.
 
-This documentation is incomplete. Contributions are welcome.
+## Contents
+- [File Structure](#file-structure)
+- [roblox](#roblox)
+- [Meta](#meta)
+- [External](#external)
+- [Item](#external)
+- [Properties](#properties)
+- [SharedStrings](#sharedstrings)
+	- [SharedString][SharedString-def] (definition)
+- [Type Elements](#type-elements)
+	- [Axes](#axes)
+	- [BinaryString](#binarystring)
+	- [bool](#bool)
+	- [BrickColor](#brickcolor)
+	- [Color3](#color3)
+	- [Color3uint8](#color3uint8)
+	- [ColorSequence](#colorsequence)
+	- [Content](#content)
+	- [CoordinateFrame](#coordinateframe)
+	- [double](#double) (Float64)
+	- [Faces](#faces)
+	- [float](#float) (Float32)
+	- [Font](#font)
+	- [int](#int) (Int32)
+	- [int64](#int64)
+	- [NumberRange](#numberrange)
+	- [NumberSequence](#numbersequence)
+	- [Optional](#optional)
+	- [PhysicalProperties](#physicalproperties)
+	- [ProtectedString](#protectedstring)
+	- [Ray](#ray)
+	- [Rect2D](#rect2d)
+	- [Ref](#ref) (Referent)
+	- [SharedString][SharedString-use] (property type)
+	- [string](#string)
+	- [token](#token) (Enum)
+	- [UDim](#udim)
+	- [UDim2](#udim2)
+	- [UniqueId](#uniqueid)
+	- [Vector2](#vector2)
+	- [Vector3](#vector3)
+	- [Vector3int16](#vector3int16)
+
+## File Structure
+Roblox XML files consist of a single `<roblox>` element, which contain a sequence of other elements. The basic layout of the file structure is as follows:
+
+- Exactly one `<roblox>` element
+	- Zero or more `<Meta>` elements
+	- Zero or more `<External>` elements
+	- One or more `<Item>` elements
+	- One `<Properties>` element
+		- Zero or more type elements
+	- Zero or more `<Item>` elements (this may nest infinitely)
+	- Zero or one `<SharedStrings>` elements
+		- Zero or more `<SharedString>` elements
+
+**To be accepted by Roblox, files MUST start with `<roblox` and end with `</roblox>`, with no whitespace.**
+
```

Comment:
The intent is to differ between suggestions and hard limitations; certain things will get a file outright rejected by Roblox or cause it to misinterpret data, and it's important to note what those are with emphasis because a lot of things are just "best practices" rather than hard limitations.

I'm open to changing this, but the current capitalization follows the IETF's recommendations in [RFC-2119](https://www.rfc-editor.org/rfc/rfc2119).

Reply:
Yes, it makes sense. There's perhaps an argument to be had about how formal these documents ought to be, but I think I'm fine with it as long as we say that we're following RFC 2119, link to it, and carefully consider each usage (particularly of SHOULD). We might also think about using the same convention in the binary spec to stay consistent.

Reply:
I'll do a pass on this to re-evaluate where we should use these in a bit. I meant to do it before finalizing the PR but evidently forgot. I know I specifically overused `SHOULD` but as I'm reading it back I misused `MAY` as well.

Wrt the binary spec, I agree and I'm planning to do a cleanup pass on it Soon:tm: because I think we can do some of it better. When I do that, I'll see where they can fit, but with binary files it starts to get weird since most of it is pretty much set in stone.

Given how we are to my knowledge the de-facto repository for knowledge on these file formats outside of Roblox (I only know of one other complete spec file for both formats), I think we should try to be as formal as possible about documenting them. People will take us at our word and I'd like to not disappoint them.